util/crc32c.cc

```diff
@@ -286,12 +285,141 @@ static const uint32_t table3_[256] = {
+and some application-specific operations:
+  - 4 String and text processing instructions
+  - 1 instruction used for comparison of packed integer quadwords
+  - 2 application-targeted accelerator (ATA) instructions:
+    - CRC32 - calculates cyclic redundancy check of a block of data
+    - POPCNT - improves searching of bit patterns
+
+
+Instruction set                         Header file
+------------------------------------------------------------------------------
+MMX                                     mmintrin.h          [Pentium MMX!]
+SSE                                     xmmintrin.h         [SSE + MMX (Pentium 3, Athlon XP)]
+SSE2                                    emmintrin.h         [SSE2 + SSE + MMX (Pentium 4, Athlon 64)]
+SSE3                                    pmmintrin.h         [SSE3 + SSE2 + SSE + MMX (Pentium 4 Prescott, Athlon 64 San Diego)]
+Suppl. SSE3                             tmmintrin.h         [SSSE3 + SSE3 + SSE2 + SSE + MMX (Core 2, Bulldozer)]
+SSE4.1                                  smmintrin.h         [SSE4_1 + SSSE3 + SSE3 + SSE2 + SSE + MMX (Core i7, Bulldozer)]
+SSE4.2                                  nmmintrin.h (MS)    [SSE4_2 + SSE4_1 + SSSE3 + SSE3 + SSE2 + SSE + MMX (Core i7, Bulldozer)]
+                                        smmintrin.h (Gnu)   [SSE4_1 + SSSE3 + SSE3 + SSE2 + SSE + MMX (Core i7, Bulldozer)]
+AES, PCLMUL                             wmmintrin.h         [AES (Core i7 Westmere, Bulldozer)]
+AVX                                     immintrin.h         [AVX, SSE4_2 + SSE4_1 + SSSE3 + SSE3 + SSE2 + SSE + MMX (Core i7 Sandy Bridge, Bulldozer)]
+POPCNT                                  popcntintrin.h      [POPCNT (Nehalem (Core i7), Phenom)]
+AMD SSE4A                               ammintrin.h         [SSE4A + SSE3 + SSE2 + SSE + MMX (Phenom)]
+AMD XOP                                 ammintrin.h (MS)    [SSE4A + SSE3 + SSE2 + SSE + MMX (Phenom)]
+                                        xopintrin.h (Gnu)   []
+AMD FMA4                                fma4intrin.h (Gnu)  []
+
+all                                     intrin.h (MS)
+                                        x86intrin.h (Gnu)   [x86 instructions]
+*/
+
+enum class CpuSSEFeature {
+/*
+  CMOV,       //("cmov")
+  MMX,        //("mmx")
+  POPCNT,     //("popcnt")
+  FMA4,       //("fma4")
+  XOP,        //("xop")
+  FMA,        //("fma")
+  AVX512F,    //("avx512f")
+  BMI,        //("bmi")
+  BMI2,       //("bmi2")
+*/
+  SSE,      //("sse")
+  SSE2,     //("sse2")
+  SSE3,     //("sse3")
+  SSSE3,    //("ssse3")
+  SSE4_A,   //("sse4a")
+  SSE4_1,   //("sse4.1")
+  SSE4_2,   //("sse4.2")
+  AVX,      //("avx")
+  AVX2,     //("avx2")
+  SSE_NONE,
+};
+
+#if !defined(__has_builtin)
+  #define __has_builtin(x) false
+#endif
+
+// We want to know if a cpu feature is available at runtime. 
+static inline bool is_cpu_feature_available(const CpuSSEFeature& cpu_feature) {
+#if (defined(__GNUC__) || (__clang__))
```

Comment:
clang defines `__GNUC__`.

Reply:
True. I spaced it when putting in the code. I will fix it.